web3/middleware/filter.py

```diff
@@ -0,0 +1,324 @@
+
+
+class RequestLogs:
+    def __init__(
+            self,
+            web3,
+            from_block=None,
+            to_block=None,
+            address=None,
+            topics=None):
+
+        self.address = address
+        self.topics = topics
+        self.web3 = web3
+        if from_block is None or from_block == 'latest':
+            self._from_block = web3.eth.blockNumber
+        else:
+            self._from_block = from_block
+        self._to_block = to_block
+        self.filter_changes = self._get_filter_changes()
+
+    @property
+    def from_block(self):
+        if self._from_block > self.web3.eth.blockNumber:
+            from_block = self.web3.eth.blockNumber
+
+        else:
+            from_block = self._from_block
+
+        return from_block
+
+    @property
+    def to_block(self):
+        if self._to_block is None:
+            to_block = self.web3.eth.blockNumber
+
+        elif self._to_block == 'latest':
+            to_block = self.web3.eth.blockNumber
+
+        elif self._to_block > self.web3.eth.blockNumber:
+            to_block = self.web3.eth.blockNumber
+
+        else:
+            to_block = self._to_block
+
+        return to_block
+
+    def _get_filter_changes(self):
+        for start, stop in iter_latest_block_ranges(self.web3, self.from_block, self._to_block):
+            if None in (start, stop):
+                yield []
+
+            yield list(
+                concat(
+                    get_logs_multipart(
+                        self.web3,
+                        start,
+                        stop,
+                        self.address,
+                        self.topics,
+                        max_blocks=5)))
+
+    def get_logs(self):
+        return list(
+            concat(
+                get_logs_multipart(
+                    self.web3,
+                    self.from_block,
+                    self.to_block,
+                    self.address,
+                    self.topics,
+                    max_blocks=5)))
+
+
+FILTER_PARAMS_KEY_MAP = {
+    'toBlock': 'to_block',
+    'fromBlock': 'from_block'
+}
+
+NEW_FILTER_METHODS = set([
+    'eth_newBlockFilter',
+    'eth_newFilter'])
+
+FILTER_CHANGES_METHODS = set([
+    'eth_getFilterChanges',
+    'eth_getFilterLogs'])
+
+
+class RequestBlocks:
+    def __init__(self, web3):
+        self.web3 = web3
+        self.start_block = web3.eth.blockNumber + 1
+
+    @property
+    def filter_changes(self):
+        return self.get_filter_changes()
+
+    def get_filter_changes(self):
+
+        block_range_iter = iter_latest_block_ranges(
+            self.web3,
+            self.start_block,
+            None)
+
+        for block_range in block_range_iter:
+            yield(block_hashes_in_range(self.web3, block_range))
+
+
+@to_list
+def block_hashes_in_range(web3, block_range):
+    for block_number in range(block_range[0], block_range[1] + 1):
+        yield web3.eth.getBlock(block_number).hash
```

Comment:
I *think* that `web3.eth.getBlock` returns `None` if `block_number` is greater than the current HEAD block number, which means this will raise an `AttributeError` if that case is met.

Reply:
That is true, I'll handle that case.